File: Examples/OneSignalDemo/app/src/main/java/com/onesignal/sdktest/notification/OneSignalNotificationSender.java
package com.onesignal.sdktest.notification;

import android.os.Build;
import android.util.Log;

import com.onesignal.OneSignal;
import com.onesignal.user.subscriptions.IPushSubscription;
import com.onesignal.sdktest.constant.Tag;
import com.onesignal.sdktest.type.Notification;

import org.json.JSONObject;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Scanner;

public class OneSignalNotificationSender {

    public static void setAppId(String appId) {
        _appId = appId;
    }
    private static String _appId = "";

    /**
     * Send a notification to this device through the OneSignal Platform.  Note this form of the
     * API should not be used in a production setting as it is not safe.  Rather, the device should
     * make an API call to it's own backend, which should handle the API call to OneSignal (where it
     * can safely provide the application API Key).
     *
     * @param notification The notification that is to be sent.
     */
    public static void sendDeviceNotification(final Notification notification) {
        new Thread(() -> {
            IPushSubscription subscription = OneSignal.getUser().getPushSubscription();

            if (!subscription.getOptedIn())
                return;

            int pos = notification.getTemplatePos();
            try {
                JSONObject notificationContent = new JSONObject("{'app_id': '" + _appId + "', 'include_player_ids': ['" + subscription.getId() + "']," +
                        "'headings': {'en': '" + notification.getTitle(pos) + "'}," +
                        "'contents': {'en': '" + notification.getMessage(pos) + "'}," +
                        "'small_icon': '" + notification.getSmallIconRes() + "'," +
                        "'large_icon': '" + notification.getLargeIconUrl(pos) + "'," +
                        "'big_picture': '" + notification.getBigPictureUrl(pos) + "'," +
                        "'android_group': '" + notification.getGroup() + "'," +
                        "'buttons': " + notification.getButtons() + "," +
                        "'android_led_color': 'FFE9444E'," +
                        "'android_accent_color': 'FFE9444E'," +
                        "'android_sound': 'nil'}");

                HttpURLConnection con = (HttpURLConnection) new URL("https://onesignal.com/api/v1/notifications").openConnection();

                con.setUseCaches(false);
                con.setConnectTimeout(30000);
                con.setReadTimeout(30000);
                con.setRequestProperty("Accept", "application/vnd.onesignal.v1+json");
                con.setRequestProperty("Content-Type", "application/json; charset=UTF-8");
                con.setRequestMethod("POST");
                con.setDoOutput(true);
                con.setDoInput(true);

                byte[] outputBytes = notificationContent.toString().getBytes(StandardCharsets.UTF_8);
                con.setFixedLengthStreamingMode(outputBytes.length);
                con.getOutputStream().write(outputBytes);

                int httpResponse = con.getResponseCode();

                if(httpResponse == HttpURLConnection.HTTP_ACCEPTED || httpResponse == HttpURLConnection.HTTP_OK) {
                    InputStream inputStream = con.getInputStream();
                    Scanner scanner = new Scanner(inputStream, "UTF-8");
                    String responseStr = "";
                    if (scanner.useDelimiter("\\A").hasNext())
                        responseStr = scanner.next();
                    scanner.close();
                    Log.d(Tag.LOG_TAG, "Success sending notification: " + responseStr);
                }
                else {
                    InputStream inputStream = con.getErrorStream();
                    Scanner scanner = new Scanner(inputStream, "UTF-8");
                    String responseStr = "";
                    if (scanner.useDelimiter("\\A").hasNext())
                        responseStr = scanner.next();
                    scanner.close();
                    Log.d(Tag.LOG_TAG, "Failure sending notification: " + responseStr);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }
}

File: Examples/OneSignalDemoV2/app/src/main/java/com/onesignal/sdktest/data/network/OneSignalNotificationSender.kt
package com.onesignal.sdktest.data.network

import android.util.Log
import com.onesignal.OneSignal
import com.onesignal.sdktest.data.model.NotificationType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Sends push notifications to the current device via the OneSignal REST API.
 * 
 * Note: This approach is for testing purposes only. In production, notifications
 * should be sent from your backend server which can securely provide the API key.
 */
object OneSignalNotificationSender {
    
    private const val TAG = "OneSignalNotificationSender"
    private const val ONESIGNAL_API_URL = "https://onesignal.com/api/v1/notifications"
    
    private var appId: String = ""
    
    fun setAppId(appId: String) {
        this.appId = appId
    }
    
    /**
     * Send a notification to this device.
     */
    suspend fun sendNotification(type: NotificationType): Boolean = withContext(Dispatchers.IO) {
        val subscription = OneSignal.User.pushSubscription
        
        if (!subscription.optedIn) {
            Log.w(TAG, "Cannot send notification - user not opted in")
            return@withContext false
        }
        
        val subscriptionId = subscription.id
        if (subscriptionId.isNullOrEmpty()) {
            Log.w(TAG, "Cannot send notification - no subscription ID")
            return@withContext false
        }
        
        try {
            val notificationJson = JSONObject().apply {
                put("app_id", appId)
                put("include_player_ids", org.json.JSONArray().put(subscriptionId))
                put("headings", JSONObject().put("en", type.notificationTitle))
                put("contents", JSONObject().put("en", type.notificationBody))
                put("android_group", type.title)
                put("android_led_color", "FFE9444E")
                put("android_accent_color", "FFE9444E")
            }
            
            val connection = (URL(ONESIGNAL_API_URL).openConnection() as HttpURLConnection).apply {
                useCaches = false
                connectTimeout = 30000
                readTimeout = 30000
                setRequestProperty("Accept", "application/vnd.onesignal.v1+json")
                setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                requestMethod = "POST"
                doOutput = true
                doInput = true
            }
            
            val outputBytes = notificationJson.toString().toByteArray(Charsets.UTF_8)
            connection.setFixedLengthStreamingMode(outputBytes.size)
            connection.outputStream.write(outputBytes)
            
            val responseCode = connection.responseCode
            
            if (responseCode == HttpURLConnection.HTTP_OK || responseCode == HttpURLConnection.HTTP_ACCEPTED) {
                val response = connection.inputStream.bufferedReader().use { it.readText() }
                Log.d(TAG, "Notification sent successfully: $response")
                return@withContext true
            } else {
                val errorResponse = connection.errorStream?.bufferedReader()?.use { it.readText() } ?: "Unknown error"
                Log.e(TAG, "Failed to send notification (HTTP $responseCode): $errorResponse")
                return@withContext false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error sending notification", e)
            return@withContext false
        }
    }
    
    /**
     * Send a custom notification with title and body.
     */
    suspend fun sendCustomNotification(title: String, body: String): Boolean = withContext(Dispatchers.IO) {
        val subscription = OneSignal.User.pushSubscription
        
        if (!subscription.optedIn) {
            Log.w(TAG, "Cannot send notification - user not opted in")
            return@withContext false
        }
        
        val subscriptionId = subscription.id
        if (subscriptionId.isNullOrEmpty()) {
            Log.w(TAG, "Cannot send notification - no subscription ID")
            return@withContext false
        }
        
        try {
            val notificationJson = JSONObject().apply {
                put("app_id", appId)
                put("include_player_ids", org.json.JSONArray().put(subscriptionId))
                put("headings", JSONObject().put("en", title))
                put("contents", JSONObject().put("en", body))
                put("android_led_color", "FFE9444E")
                put("android_accent_color", "FFE9444E")
            }
            
            val connection = (URL(ONESIGNAL_API_URL).openConnection() as HttpURLConnection).apply {
                useCaches = false
                connectTimeout = 30000
                readTimeout = 30000
                setRequestProperty("Accept", "application/vnd.onesignal.v1+json")
                setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                requestMethod = "POST"
                doOutput = true
                doInput = true
            }
            
            val outputBytes = notificationJson.toString().toByteArray(Charsets.UTF_8)
            connection.setFixedLengthStreamingMode(outputBytes.size)
            connection.outputStream.write(outputBytes)
            
            val responseCode = connection.responseCode
            
            if (responseCode == HttpURLConnection.HTTP_OK || responseCode == HttpURLConnection.HTTP_ACCEPTED) {
                val response = connection.inputStream.bufferedReader().use { it.readText() }
                Log.d(TAG, "Custom notification sent successfully: $response")
                return@withContext true
            } else {
                val errorResponse = connection.errorStream?.bufferedReader()?.use { it.readText() } ?: "Unknown error"
                Log.e(TAG, "Failed to send custom notification (HTTP $responseCode): $errorResponse")
                return@withContext false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error sending custom notification", e)
            return@withContext false
        }
    }
}
